src/JuMPContainer.jl

```diff
@@ -183,13 +185,86 @@ Base.length(it::ValueIterator)  = length(it.x)
 
 type KeyIterator{JA<:JuMPArray}
     x::JA
+    dim::Int
+    next_k_cache::Array{Any,1}
+    function KeyIterator(d)
+        n = ndims(d.innerArray)
+        new(d, n, Array(Any, n+1))
+    end
+end
+
+KeyIterator{JA}(d::JA) = KeyIterator{JA}(d)
+
+immutable IndexableIndexsets end
+immutable NotIndexableIndexsets end
+
+function indexability(x::JuMPArray)
+    allsetsindexable = true
+    for i in  1:length(x.indexsets)
+        if !method_exists(getindex, (typeof(x.indexsets[i]),))
+            allsetsindexable = false
+            break
+        end
+    end
+
+    if allsetsindexable
+        IndexableIndexsets()
+    else
+        NotIndexableIndexsets()
+    end
+end
+
+Base.start(it::KeyIterator) = _start(it, indexability(it.x))
+_start(it::KeyIterator, ::IndexableIndexsets) = start(it.x.innerArray)
+_start(it::KeyIterator, ::NotIndexableIndexsets) = notindexable_start(it.x)
+
+@generated function notindexable_start{T,N,NT}(x::JuMPArray{T,N,NT})
+    quote
+        $(Expr(:tuple, 0, [:(start(x.indexsets[$i])) for i in 1:N]...))
+    end
 end
-Base.start(it::KeyIterator)   =  start(it.x.innerArray)
-@generated __next{T,N,NT}(x::JuMPArray{T,N,NT}, k) =
+
+@generated function _next{T,N,NT}(x::JuMPArray{T,N,NT}, k::Tuple)
+    quote
+        $(Expr(:tuple, [:(next(x.indexsets[$i], k[$i+1])[1]) for i in 1:N]...))
+    end
+end
+
+function Base.next(it::KeyIterator, k::Tuple)
+    cartesian_key = JuMPKey(_next(it.x, k))
+    pos = -1
+    for i in 1:it.dim
+        if(!done(it.x.indexsets[i], next(it.x.indexsets[i], k[i+1])[2] ) )
+            pos = i
+            break
+        end
+    end
+    if pos == - 1
+        return cartesian_key, (1,1) #The second 1 is needed to ambiguity with ints
```

Comment:
The first element of the tuple is used for indicating that we are done. The second 1 here is added because otherwise the dispatch of the function `next` fails and the function `next` with `Integer` state is called instead of the one with `Tuple` state